File: crates/spfs/src/cli/_cmd_platforms.py
import argparse

import spfs


def register(sub_parsers: argparse._SubParsersAction) -> None:

    platforms_cmd = sub_parsers.add_parser("platforms", help=_platforms.__doc__)
    platforms_cmd.add_argument(
        "--remote",
        "-r",
        help="Show platforms from remote repository instead of the local one",
    )
    platforms_cmd.set_defaults(func=_platforms)


def _platforms(args: argparse.Namespace) -> None:
    """List all platforms in an spfs repository."""

    config = spfs.get_config()
    if args.remote is not None:
        repo = config.get_remote(args.remote)
    else:
        repo = config.get_repository()

    platforms = repo.iter_platforms()
    for platform in platforms:
        print(spfs.io.format_digest(platform.digest()))

File: crates/spfs/src/cli/cmd_platforms.rs
// Copyright (c) 2021 Sony Pictures Imageworks, et al.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/imageworks/spk

use clap::Args;
use tokio_stream::StreamExt;

use spfs::io::{self, DigestFormat};

/// List all platforms in an spfs repository
#[derive(Debug, Args)]
pub struct CmdPlatforms {
    /// Show layers from remote repository instead of the local one
    #[clap(long, short)]
    remote: Option<String>,

    /// Show the shortened form of each reported layer digest
    #[clap(long)]
    short: bool,

    /// Also find and report any tags that point to each platform, implies --short
    #[clap(long)]
    tags: bool,
}

impl CmdPlatforms {
    pub async fn run(&mut self, config: &spfs::Config) -> spfs::Result<i32> {
        let repo = match &self.remote {
            Some(remote) => config.get_remote(remote).await?,
            None => config.get_repository().await?.into(),
        };
        let mut platforms = repo.iter_platforms();
        while let Some(platform) = platforms.next().await {
            let (digest, _) = platform?;
            println!("{}", self.format_digest(digest, &repo).await?);
        }
        Ok(0)
    }

    async fn format_digest<'repo>(
        &self,
        digest: spfs::encoding::Digest,
        repo: &'repo spfs::storage::RepositoryHandle,
    ) -> spfs::Result<String> {
        if self.tags {
            io::format_digest(digest, DigestFormat::ShortenedWithTags(repo)).await
        } else if self.short {
            io::format_digest(digest, DigestFormat::Shortened(repo)).await
        } else {
            io::format_digest(digest, DigestFormat::Full).await
        }
    }
}
